core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -764,4 +762,62 @@ public String getStderr() {
             return stderr;
         }
     }
+
+    /**
+     * Convenience class with access to non-public members of GenericContainer.
+     */
+    public static abstract class AbstractWaitStrategy implements WaitStrategy {
+        protected GenericContainer container;
+
+        @NonNull
+        protected Duration startupTimeout = Duration.ofSeconds(60);
+
+        /**
+         * Set the container for which to wait
+         * @param container the container
+         * @return this
+         */
+        @Override
+        public WaitStrategy forContainer(GenericContainer container) {
+            this.container = container;
+            return this;
+        }
+
+        /**
+         * Set the duration of waiting time until container treated as started.
+         *
+         * @param startupTimeout timeout
+         * @return this
+         * @see WaitStrategy#waitUntilReady()
+         */
+        public WaitStrategy withStartupTimeout(Duration startupTimeout) {
+            this.startupTimeout = startupTimeout;
+            return this;
+        }
+
+        /**
+         * @return the container's logger
+         */
+        protected Logger logger() {
+            return container.logger();
+        }
+
+        /**
+         * @return the port on which to check if the container is ready
+         */
+        protected Integer getLivenessCheckPort() {
+            return ofNullable(container.getLivenessCheckPort())
+                    .orElse(container.getPortBindings().stream()
+                            .map(pb -> PortBinding.parse(pb).getBinding().getHostPort())
+                            .findAny()
+                            .orElse(null));
+        }
```

Comment:
See above - I think this could maybe be a bit more readable by taking a less clever approach!